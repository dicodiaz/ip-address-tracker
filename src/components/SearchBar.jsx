import { useEffect, useState } from 'react';
import { Button, Col, Container, Form, Image, Row } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import iconArrow from '../assets/icon-arrow.svg';
import CardInfo from './CardInfo';
import { selectGeolocation } from '../redux/store';
import { fetchGeolocationFromIpAddress } from '../redux/slices/geolocation-slice';

const SearchBar = () => {
  const dispatch = useDispatch();
  const geolocation = useSelector(selectGeolocation);
  const [inputValue, setInputValue] = useState('');
  const [submitValue, setSubmitValue] = useState('8.8.8.8');

  useEffect(() => {
    dispatch(fetchGeolocationFromIpAddress(submitValue));
  }, [dispatch, submitValue]);

  const handleSubmit = (e) => {
    e.preventDefault();
    setSubmitValue(inputValue);
    setInputValue('');
  };

  return (
    <Container className="px-0 flex-grow-1 bg-search-bar d-flex flex-column" as="section" fluid>
      <h1 className="text-center text-white mt-4 mb-0">IP Address Tracker</h1>
      <Form className="mt-4" onSubmit={handleSubmit}>
        <Row className="mx-0 g-0 justify-content-center" xs="auto">
          <Col xs={10} md={6}>
            <Form.Control
              className="form-control-border-radius"
              placeholder="Search for any IP address or domain"
              value={inputValue}
              onChange={(e) => setInputValue(e.currentTarget.value)}
              size="lg"
            />
          </Col>
          <Col>
            <Button className="button-border-radius" type="submit" variant="dark" size="lg">
              <Image className="align-baseline" src={iconArrow} />
            </Button>
          </Col>
        </Row>
      </Form>
      <CardInfo geolocation={geolocation} />
    </Container>
  );
};

export default SearchBar;
