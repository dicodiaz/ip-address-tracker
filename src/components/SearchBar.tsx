import { ChangeEvent, FC, useEffect, useState } from 'react';
import { Button, Col, Container, Form, Image, Row } from 'react-bootstrap';
import iconArrow from '../assets/icon-arrow.svg';
import { useAppDispatch } from '../hooks/redux-hooks';
import { fetchGeolocationFromIpAddress } from '../redux/slices/geolocationSlice';

const SearchBar: FC = () => {
  const dispatch = useAppDispatch();
  const [inputValue, setInputValue] = useState('');
  const [submitValue, setSubmitValue] = useState('8.8.8.8');
  const [isInvalid, setIsInvalid] = useState(false);

  useEffect(() => {
    dispatch(fetchGeolocationFromIpAddress(submitValue));
  }, [dispatch, submitValue]);

  const handleSubmit = (e: ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!/^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$/.test(inputValue)) {
      setIsInvalid(true);
      return;
    }
    setSubmitValue(inputValue);
    setInputValue('');
    setIsInvalid(false);
  };

  return (
    <Container as="section" className="bg-search-bar flex-grow-1 py-4" fluid>
      <h1 className="text-white text-center mb-3">IP Address Tracker</h1>
      <Form onSubmit={handleSubmit}>
        <Row className="mx-0 g-0 justify-content-center">
          <Col className="mw-form-control">
            <Form.Control
              placeholder="Search for any IP address"
              value={inputValue}
              onChange={(e) => setInputValue(e.currentTarget.value)}
              size="lg"
              isInvalid={isInvalid}
              aria-label="search-ip"
            />
            <Form.Control.Feedback type="invalid">
              Please enter valid IP address
            </Form.Control.Feedback>
          </Col>
          <Col xs="auto">
            <Button type="submit" variant="dark" size="lg" aria-label="submit">
              <Image className="align-baseline" src={iconArrow} aria-label="icon-arrow" />
            </Button>
          </Col>
        </Row>
      </Form>
    </Container>
  );
};

export default SearchBar;
