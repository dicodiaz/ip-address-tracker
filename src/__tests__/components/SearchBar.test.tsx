import userEvent from '@testing-library/user-event';
import axios from 'axios';
import { expect } from 'vitest';
import SearchBar from '../../components/SearchBar';
import { renderWithProviders } from '../test-utils';

describe('SearchBar component', () => {
  it('should render as expected', () => {
    const { container } = renderWithProviders(<SearchBar />);
    expect(container).toMatchSnapshot();
  });

  it('should fetch once on initial rendering', async () => {
    renderWithProviders(<SearchBar />);
    expect(axios.get).toHaveBeenCalledTimes(1);
  });

  it('should execute an additional fetch when submitting a valid IP', async () => {
    const { getByRole } = renderWithProviders(<SearchBar />);
    await userEvent.type(getByRole('textbox', { name: 'search-ip' }), '1.1.1.1');
    await userEvent.click(getByRole('button', { name: 'submit' }));
    expect(axios.get).toHaveBeenCalledTimes(2);
  });

  it('should execute an additional fetch when submitting an invalid IP', async () => {
    const { getByRole } = renderWithProviders(<SearchBar />);
    await userEvent.type(getByRole('textbox', { name: 'search-ip' }), '1.1.1');
    await userEvent.click(getByRole('button', { name: 'submit' }));
    expect(axios.get).toHaveBeenCalledTimes(1);
  });
});
