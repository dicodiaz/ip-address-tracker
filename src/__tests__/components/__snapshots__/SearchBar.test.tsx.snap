// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar component > should render as expected 1`] = `
<div>
  <section
    class="bg-search-bar flex-grow-1 py-4 container-fluid"
  >
    <h1
      class="text-white text-center mb-3"
    >
      IP Address Tracker
    </h1>
    <form
      class=""
    >
      <div
        class="mx-0 g-0 justify-content-center row"
      >
        <div
          class="mw-form-control col"
        >
          <input
            aria-label="search-ip"
            class="form-control form-control-lg"
            placeholder="Search for any IP address"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Please enter valid IP address
          </div>
        </div>
        <div
          class="col-auto"
        >
          <button
            aria-label="submit"
            class="btn btn-dark btn-lg"
            type="submit"
          >
            <img
              aria-label="icon-arrow"
              class="align-baseline"
              src="/src/assets/icon-arrow.svg"
            />
          </button>
        </div>
      </div>
    </form>
  </section>
</div>
`;
